src/pydrex/core.py

```diff
@@ -0,0 +1,634 @@
+            orientations[element_index],
+            slip_rates,
+            rrss,
+            velocity_gradient,
+        )
+
+        # Spin vector for the volume element, see eq. 3 in Fraters 2021.
+        spin_vector = np.empty(3)
+        for j in range(3):
+            r = (j + 1) % 3
+            s = (j + 2) % 3
+            spin_vector[j] = (
+                velocity_gradient[s, r]
+                - velocity_gradient[r, s]
+                - (deformation_rate[s, r] - deformation_rate[r, s]) * slip_rate_softest
+            ) / 2
+
+        # Calculate rotation rate, see eq. 9 Kaminski & Ribe (2001).
+        for p in range(3):
+            for q in range(3):
+                for r in range(3):
+                    for s in range(3):
+                        rotation_rates[element_index, p, q] += (
+                            _tensors.PERMUTATION_SYMBOL[q, r, s]
+                            * orientations[element_index, p, s]
+                            * spin_vector[r]
+                        )
+
+    return rotation_rates, slip_rates, slip_indices, slip_rate_softest
+
+
+@nb.njit(fastmath=True)
+def get_strain_energy_residuals(
+    vol_dist,
+    slip_rates,
+    slip_indices,
+    slip_rate_softest,
+    rrss,
+    stress_exponent,
+    nucleation_efficiency,
+):
+    """Calculate strain energy residual for each volume element.
+
+    Args:
+        `vol_dist` (NxD array) — present volume distribution (discretized)
+        `slip_rates` (array) — slip rates relative to slip rate on softest slip system
+        `slip_indices` (array) — indices that sort the RRSS by increasing slip rate activity
+        `slip_rate_softest` (float) — slip rate on the softest (most active) slip system
+        `stress_exponent` (float) — exponent for the stress dependence of dislocation density
+        `nucleation_efficiency` (float) — parameter controlling grain nucleation
+
+    """
+    n_elements = len(vol_dist)
+    strain_energies = np.zeros(n_elements)
+    for element_index in range(n_elements):
+        # Dimensionless dislocation density for each slip system.
+        # See eq. 16 Fraters 2021.
+        # NOTE: Mistake in eq. 11, Kaminski 2004: spurrious division by strain rate scale.
+        if rrss == _fabric.RRSS_OLIVINE_A:
+            for i in slip_indices[1:]:
+                # TODO: Verify rrss[i] == τ_0 / τ^sv
+                dislocation_density = rrss[i] ** (1.5 - stress_exponent) * np.abs(
+                    slip_rates[i] * slip_rate_softest
+                ) ** (1.5 / stress_exponent)
+                # Dimensionless strain energy for this element, see eq. 14, Fraters 2021.
+                strain_energies[element_index] += dislocation_density * np.exp(
+                    -nucleation_efficiency * dislocation_density ** 2
+                )
+        elif rrss == _fabric.RRSS_ENSTATITE:
+            # NOTE: Mistake in old PyDRex: x / y ** a != (x / y) ** a
+            # See original DRex code, line 1437.
```

Comment:
Not sure I see the mistake.

Reply:
I don't know why I thought it was incorrect: https://github.com/Patol75/PyDRex/blob/master/src/PyDRex.py#L369

but actually it's the same because of `(1/x)**a == 1/x**a`. I'll remove the note.